Câu 1: 
Để lập mô hình hồi quy tuyến tính mô tả mối quan hệ giữa tổng vốn đầu tư và lãi suất ngân hàng, ta sử dụng mô hình hồi quy đơn (simple linear regression model) với biến độc lập là tổng vốn đầu tư (X) và biến phụ thuộc là lãi suất ngân hàng (Y).

Dựa trên đoạn code trước đó, ta đã xây dựng mô hình hồi quy tuyến tính và thu được kết quả. Để mô tả mô hình, ta có các thông số sau:

Hệ số hồi quy (coefficients): Trong trường hợp này, hệ số hồi quy đơn giữa tổng vốn đầu tư (X) và lãi suất ngân hàng (Y) được ký hiệu là results.params[1], có giá trị là -9.5323. Điều này cho biết mỗi đơn vị tăng (giảm) trong tổng vốn đầu tư thì lãi suất ngân hàng sẽ giảm (tăng) 9.5323 đơn vị.

Ý nghĩa của hệ số hồi quy: Ý nghĩa của hệ số hồi quy trong trường hợp này là cho biết mức độ thay đổi của lãi suất ngân hàng khi tổng vốn đầu tư thay đổi. Hệ số hồi quy âm (-9.5323) cho thấy mối quan hệ nghịch đảo giữa tổng vốn đầu tư và lãi suất ngân hàng, tức là khi tổng vốn đầu tư tăng, lãi suất ngân hàng sẽ giảm và ngược lại.

R-squared (R^2): Trong kết quả mô hình, R-squared được đưa ra là 0.920, có nghĩa là mô hình giải thích được 92% sự biến động của lãi suất ngân hàng dựa trên tổng vốn đầu tư. Điều này cho thấy mô hình có mức độ phù hợp tương đối cao trong việc giải thích biến động của lãi suất ngân hàng dựa trên tổng vốn đầu tư.

Adjusted R-squared (Adj. R^2): Được tính toán để điều chỉnh cho số lượng biến động dẫn đến mất mát tự do (degrees of freedom). Giá trị này điều chỉnh R-squared để đánh giá mức độ phù hợp của mô hình trong việc giải thích biến động của dữ liệu, đồng thời xem xét số lượng biến động và số lượng quan sát.








In [ ]:
import statsmodels.api as sm
import numpy as np
from sklearn.utils import resample

# Câu 1: Xây dựng mô hình hồi quy tuyến tính
X = np.array([7.0, 6.5, 6.5, 6.0, 6.0, 6.0, 5.5, 5.5, 5.0, 4.5])
Y = np.array([29, 32, 31, 34, 35, 32, 40, 43, 48, 50])
X = sm.add_constant(X)  # Thêm hằng số vào X

model = sm.OLS(Y, X)  
results = model.fit()  # Tiến hành phân tích
# In kết quả mô hình
print("\nCâu 1: Xây dựng mô hình hồi quy tuyến tính")
print(results.summary())



Câu 1: Xây dựng mô hình hồi quy tuyến tính
                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.920
Model:                            OLS   Adj. R-squared:                  0.910
Method:                 Least Squares   F-statistic:                     91.78
Date:                Fri, 26 May 2023   Prob (F-statistic):           1.17e-05
Time:                        02:28:05   Log-Likelihood:                -21.096
No. Observations:                  10   AIC:                             46.19
Df Residuals:                       8   BIC:                             46.80
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const   

/usr/local/lib/python3.10/dist-packages/scipy/stats/_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


Câu 2: Dựa trên kết quả mô tả mô hình hồi quy tuyến tính, chúng ta có thể rút ra các kết luận sau:

Ý nghĩa của các hệ số hồi quy:

Hệ số hồi quy const (hằng số): Có giá trị là 93.1642 và có ý nghĩa thống kê với p-value rất nhỏ (0.000). Điều này cho biết tổng vốn đầu tư có ảnh hưởng đáng kể đến biến phụ thuộc (lãi suất ngân hàng).

Hệ số hồi quy x1 (lãi suất ngân hàng): Có giá trị là -9.5323 và cũng có ý nghĩa thống kê với p-value rất nhỏ (0.000). Điều này cho biết lãi suất ngân hàng có ảnh hưởng đáng kể đến tổng vốn đầu tư.
Đánh giá mức độ phù hợp của mô hình:

R-squared (R^2): Có giá trị là 0.920, cho thấy mô hình hồi quy tuyến tính giải thích được khoảng 92% sự biến thiên của biến phụ thuộc (tổng vốn đầu tư). R^2 càng gần 1 thì mô hình càng phù hợp.

F-statistic: Có giá trị là 91.78 và p-value rất nhỏ (1.17e-05), cho thấy mô hình có sự ảnh hưởng đáng kể đến biến phụ thuộc. Điều này chỉ ra rằng tổng vốn đầu tư và lãi suất ngân hàng có mối quan hệ đáng kể.

Tổng quan, mô hình hồi quy tuyến tính đơn giữa tổng vốn đầu tư và lãi suất ngân hàng có ý nghĩa thống kê và mức độ phù hợp tương đối cao. Tuy nhiên, cần lưu ý rằng số lượng quan sát trong mô hình chỉ là 10, do đó kết quả cần được xem xét thận trọng và không áp dụng tổng quát cho quần thể lớn hơn.






In [ ]:
# Câu 2: Ý nghĩa của các hệ số hồi quy
print("\nCâu 2: Ý nghĩa của các hệ số hồi quy")
print(f"Hệ số hồi quy 0 (hằng số): {results.params[0]}")
print(f"Hệ số hồi quy 1: {results.params[1]}")


Câu 2: Ý nghĩa của các hệ số hồi quy
Hệ số hồi quy 0 (hằng số): 93.16417910447765
Hệ số hồi quy 1: -9.532338308457724


Câu 3: 
Cách kiểm định sau sử dụng phương pháp bootstrap để tạo ra các mẫu tái chọn (bootstrap samples) từ mẫu gốc. Cụ thể, các mẫu bootstrap được lấy từ mẫu gốc bằng cách lấy ngẫu nhiên các quan sát có hoàn lại. Sau đó, mô hình hồi quy tuyến tính được đào tạo trên mỗi mẫu bootstrap để tính toán hệ số hồi quy.

Cách kiểm định đầu tiên trong đoạn code ban đầu sử dụng các phép tính thống kê từ thư viện scipy.stats để tính toán giá trị t-statistic và p-value dựa trên mô hình hồi quy tuyến tính ban đầu. Tuy nhiên, phương pháp này có một giả định rằng mẫu phải có số lượng quan sát lớn hơn hoặc bằng 20 để kết quả kiểm định có ý nghĩa. Trong trường hợp mẫu nhỏ hơn 20 (như trong ví dụ của bạn với n = 10), kết quả kiểm định có thể không chính xác hoặc không đáng tin cậy.

Do đó, cách kiểm định sau sử dụng phương pháp bootstrap để tạo ra các mẫu tái chọn và đào tạo mô hình hồi quy trên mỗi mẫu bootstrap để tính toán giá trị dự báo và khoảng tin cậy. Phương pháp bootstrap không có giả định về kích thước mẫu và có thể áp dụng cho các mẫu nhỏ.






In [ ]:
# Câu 3: Kiểm định giả thiết "Hệ số hồi quy của X bằng 0 với mức ý nghĩa 2%" sử dụng bootstrapping
hypothesis = "X = 0"

n_bootstrap = 1000  # Số lần lấy mẫu bootstrap
bootstrap_coefficients = []
for _ in range(n_bootstrap):
    bootstrap_sample = resample(X, replace=True)  # Lấy mẫu theo phương pháp bootstrap
    bootstrap_X = bootstrap_sample[:, :2]  # Lấy X từ mẫu bootstrap
    bootstrap_Y = bootstrap_sample[:, :1]  # Lấy Y từ mẫu bootstrap
    bootstrap_model = sm.OLS(bootstrap_Y, bootstrap_X)  # Xây dựng mô hình bootstrap
    bootstrap_results = bootstrap_model.fit()  # Đào tạo mô hình bootstrap
    bootstrap_coefficients.append(bootstrap_results.params[1])  # Lưu trữ hệ số hồi quy của X

# Tính p-value
observed_coefficient = results.params[1]  # Hệ số hồi quy quan sát
p_value = (np.abs(bootstrap_coefficients) >= np.abs(observed_coefficient)).mean()

# In kết quả kiểm định
print("Kiểm định giả thiết")
print(f"Giả thiết: {hypothesis}")
print(f"Giá trị quan sát t: {observed_coefficient}")
print(f"Giá trị p-value: {p_value}")
alpha = 0.02
if p_value < alpha:
    print("\nCâu 3: Kiểm định giả thiết Hệ số hồi quy của X bằng 0 với mức ý nghĩa 2%")
    print("Kết luận: Có đủ bằng chứng để bác bỏ giả thiết")
else:
    print("Kết luận: Không đủ bằng chứng để bác bỏ giả thiết")


Kiểm định giả thiết
Giả thiết: X = 0
Giá trị quan sát t: -9.532338308457724
Giá trị p-value: 0.0

Câu 3: Kiểm định giả thiết Hệ số hồi quy của X bằng 0 với mức ý nghĩa 2%
Kết luận: Có đủ bằng chứng để bác bỏ giả thiết


In [ ]:
# Câu 3: Kiểm định giả thiết "Hệ số hồi quy của X bằng 0 với mức ý nghĩa 2%"
hypothesis = "X = 0"
t_statistic = results.tvalues[1]
p_value = results.pvalues[1]

# In kết quả kiểm định
print("Kiểm định giả thiết")
print(f"Giả thiết: {hypothesis}")
print(f"Giá trị t-statistic: {t_statistic}")
print(f"Giá trị p-value: {p_value}")
alpha = 0.02
if p_value < alpha:
    print("\nCâu 3: Kiểm định giả thiết Hệ số hồi quy của X bằng 0 với mức ý nghĩa 2%")
    print("Kết luận: Có đủ bằng chứng để bác bỏ giả thiết")
else:
    print("Kết luận: Không đủ bằng chứng để bác bỏ giả thiết")


Kiểm định giả thiết
Giả thiết: X = 0
Giá trị t-statistic: -9.580000000000005
Giá trị p-value: 1.1682439930331031e-05

Câu 3: Kiểm định giả thiết Hệ số hồi quy của X bằng 0 với mức ý nghĩa 2%
Kết luận: Có đủ bằng chứng để bác bỏ giả thiết


In [ ]:
# Câu 4: Dự báo tổng vốn đầu tư trung bình khi lãi suất là 4.8% năm với độ tin cậy 98%
new_X = np.array([1, 4.8])  # Giá trị mới cho X (bao gồm hằng số 1)

# Tạo mẫu bootstrap
n_bootstrap = 1000  # Số lần lấy mẫu
bootstrap_predictions = []
for _ in range(n_bootstrap):
    bootstrap_sample = resample(X, replace=True)  # Lấy mẫu theo phương pháp bootstrap
    bootstrap_X = bootstrap_sample[:, :2]  # Lấy X từ mẫu bootstrap
    bootstrap_Y = bootstrap_sample[:, :1]  # Lấy Y từ mẫu bootstrap
    bootstrap_model = sm.OLS(bootstrap_Y, bootstrap_X)  # Xây dựng mô hình bootstrap
    bootstrap_results = bootstrap_model.fit()  # Đào tạo mô hình bootstrap
    bootstrap_prediction = np.dot(new_X, bootstrap_results.params)  # Dự đoán Y trên mẫu bootstrap
    bootstrap_predictions.append(bootstrap_prediction)

# Tính khoảng tin cậy
alpha = 0.02
# Tính khoảng tin cậy
lower_bound = np.percentile(bootstrap_predictions, alpha/2 * 100)
upper_bound = np.percentile(bootstrap_predictions, (1 - alpha/2) * 100)

# In kết quả dự báo
print("\nCâu 4: Dự báo tổng vốn đầu tư trung bình")
print(f"Giá trị dự báo: {np.mean(bootstrap_predictions)}")
print(f"Khoảng tin cậy {100*(1-alpha)}%: ({lower_bound}, {upper_bound})")


Câu 4: Dự báo tổng vốn đầu tư trung bình
Giá trị dự báo: 0.9999999999999998
Khoảng tin cậy 98.0%: (0.9999999999999988, 1.0000000000000009)
